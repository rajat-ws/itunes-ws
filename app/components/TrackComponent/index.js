/**
 *
 * TrackComponent
 *
 */

import React, { memo, useRef, useState } from 'react';
import { useHistory } from 'react-router-dom';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Card, Image, Typography } from 'antd';
import { isEmpty } from 'lodash';
import { colors, fonts } from '@app/themes/index';
import If from '@components/If';

const { Paragraph } = Typography;

const TrackCardContainer = styled(Card)`
  && {
    border-radius: 0.5rem;
    width: 25rem;
    border: 1px solid ${colors.secondary};
    text-align: center;
`;

const StyledDescription = styled.div`
  display: flex;
  flex-direction: column;
`;

const StyledSpan = styled.span`
  && {
    color: ${colors.background};
    ${fonts.weights.bold}
    ${fonts.size.big}
  }
`;

const PlayTrackBtn = styled.button`
  && {
    background-color: ${colors.background};
    border: none;
    display: flex;
    justify-content: center;
    align-items: center;
    border-radius: 0.5rem;
    width: 100%;
    cursor: pointer;

    &:hover {
      background-color: ${colors.primaryLight};
      color: ${colors.text};
    }
  }
`;

const ShowDetailsBtn = styled.button`
  && {
    background-color: ${colors.primaryDark};
    border: none;
    display: flex;
    justify-content: center;
    align-items: center;
    border-radius: 0.5rem;
    width: 100%;
    cursor: pointer;

    &:hover {
      background-color: ${colors.primaryLight};
      color: ${colors.text};
    }
  }
`;

const ButtonWrapper = styled.div`
  && {
    display: flex;
    gap: 1rem;
  }
`;

const ButtonLabel = styled.span`
  && {
    color: ${colors.text};
    ${fonts.weights.bold}
    ${fonts.size.big}
  }
`;

export function TrackComponent({ trackData, isShowDetailsButton, isShowDetails, handlePauseTrackWrapper }) {
  const [isTrackPlaying, setIsTrackPlaying] = useState(false);

  const {
    artistName,
    artworkUrl100: imageUrl,
    previewUrl,
    trackTimeMillis: trackDuration,
    collectionName,
    trackName,
    trackId,
    country,
    primaryGenreName,
    kind,
    wrapperType
  } = trackData;
  const audioRef = useRef(null);
  const history = useHistory();

  const handlePlayPauseBtn = e => {
    e.preventDefault();

    const isTrackPaused = audioRef.current?.paused;

    if (isTrackPaused) {
      audioRef.current.play();
    } else {
      audioRef.current.pause();
    }
    setIsTrackPlaying(!isTrackPlaying);
    handlePauseTrackWrapper(audioRef);
  };

  const handleTrackDetailsRoute = trackId => history.push(`/tracks/${trackId}`);

  return (
    <TrackCardContainer data-testid="track-component">
      <If condition={!isEmpty(imageUrl)} otherwise={<Image>No image available</Image>}>
        <Image src={imageUrl} width="80%" />
      </If>

      <StyledDescription>
        <If condition={!isEmpty(artistName)} otherwise={<Paragraph>No artist name available</Paragraph>}>
          <Paragraph>
            <StyledSpan> Artist name: </StyledSpan> {artistName}
          </Paragraph>
        </If>

        <If condition={!isEmpty(collectionName)} otherwise={<Paragraph>No collection name available</Paragraph>}>
          <Paragraph>
            <StyledSpan> Collection name: </StyledSpan> {collectionName}
          </Paragraph>
        </If>

        <If condition={!isEmpty(trackName)} otherwise={<Paragraph>No track name available</Paragraph>}>
          <Paragraph>
            <StyledSpan> Track name: </StyledSpan> {trackName}
          </Paragraph>
        </If>
        <If condition={isShowDetails}>
          <Paragraph>
            <StyledSpan> Country: </StyledSpan> {country}
          </Paragraph>
          <Paragraph>
            <StyledSpan> Kind: </StyledSpan> {kind}
          </Paragraph>
          <Paragraph>
            <StyledSpan> Genre: </StyledSpan> {primaryGenreName}
          </Paragraph>
          <Paragraph>
            <StyledSpan> Wrapper Type: </StyledSpan> {wrapperType}
          </Paragraph>

          {trackDuration && (
            <Paragraph>
              <StyledSpan> Duration: </StyledSpan> {Math.floor(trackDuration / 60000)}:
              {Math.floor(trackDuration / 1000) % 60}s
            </Paragraph>
          )}
        </If>
      </StyledDescription>

      <ButtonWrapper>
        <If condition={isShowDetailsButton}>
          <ShowDetailsBtn onClick={() => handleTrackDetailsRoute(trackId)}>
            <ButtonLabel> Show Details </ButtonLabel>
          </ShowDetailsBtn>
        </If>

        <PlayTrackBtn onClick={handlePlayPauseBtn}>
          <If
            condition={!audioRef.current?.paused && audioRef.current?.src}
            otherwise={<ButtonLabel> Play </ButtonLabel>}
          >
            <ButtonLabel> Pause </ButtonLabel>
          </If>
        </PlayTrackBtn>
      </ButtonWrapper>
      <audio src={previewUrl} ref={audioRef} data-testid="trackAudio" />
    </TrackCardContainer>
  );
}

export default memo(TrackComponent);

TrackComponent.defaultProps = {
  trackData: {}
};

TrackComponent.propTypes = {
  trackData: PropTypes.shape({
    artistId: PropTypes.number,
    trackTimeMillis: PropTypes.number,
    trackId: PropTypes.number,
    artistName: PropTypes.string,
    artworkUrl100: PropTypes.string,
    collectionName: PropTypes.string,
    trackName: PropTypes.string,
    maxWidth: PropTypes.number,
    songId: PropTypes.number,
    imageUrl: PropTypes.string,
    wrapperType: PropTypes.string,
    primaryGenreName: PropTypes.string,
    kind: PropTypes.string,
    country: PropTypes.string,
    previewUrl: PropTypes.string
  }),

  isShowDetailsButton: PropTypes.bool,
  isShowDetails: PropTypes.bool,
  handlePauseTrackWrapper: PropTypes.func
};
