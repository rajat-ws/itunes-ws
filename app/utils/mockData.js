export const MOCK_TRACK_DATA = {
  wrapperType: 'track',
  kind: 'song',
  artistId: 76531581,
  collectionId: 1068455603,
  trackId: 1068455608,
  artistName: 'Tom MacDonald',
  collectionName: 'See You Tomorrow',
  trackName: 'Greenday',
  collectionCensoredName: 'See You Tomorrow',
  trackCensoredName: 'Greenday',
  artistViewUrl: 'https://music.apple.com/us/artist/tom-macdonald/76531581?uo=4',
  collectionViewUrl: 'https://music.apple.com/us/album/greenday/1068455603?i=1068455608&uo=4',
  trackViewUrl: 'https://music.apple.com/us/album/greenday/1068455603?i=1068455608&uo=4',
  previewUrl:
    'https://audio-ssl.itunes.apple.com/itunes-assets/AudioPreview125/v4/51/bb/61/51bb6190-6865-3050-af01-0f64502bc376/mzaf_14313183610807291436.plus.aac.p.m4a',
  artworkUrl30:
    'https://is3-ssl.mzstatic.com/image/thumb/Music124/v4/dd/64/78/dd6478bc-e684-1b69-2e4e-c7e1d6a36c0a/045635953260.jpg/30x30bb.jpg',
  artworkUrl60:
    'https://is3-ssl.mzstatic.com/image/thumb/Music124/v4/dd/64/78/dd6478bc-e684-1b69-2e4e-c7e1d6a36c0a/045635953260.jpg/60x60bb.jpg',
  artworkUrl100:
    'https://is3-ssl.mzstatic.com/image/thumb/Music124/v4/dd/64/78/dd6478bc-e684-1b69-2e4e-c7e1d6a36c0a/045635953260.jpg/100x100bb.jpg',
  collectionPrice: 9.99,
  trackPrice: 0.99,
  releaseDate: '2015-12-21T12:00:00Z',
  collectionExplicitness: 'explicit',
  trackExplicitness: 'explicit',
  discCount: 1,
  discNumber: 1,
  trackCount: 17,
  trackNumber: 3,
  trackTimeMillis: 256520,
  country: 'USA',
  currency: 'USD',
  primaryGenreName: 'Hip-Hop/Rap',
  contentAdvisoryRating: 'Explicit',
  isStreamable: true
};
