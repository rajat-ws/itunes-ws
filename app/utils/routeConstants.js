export default {
  itunes: {
    route: '/',
    props: {
      maxWidth: 32,
      padding: 20
    },
    exact: true
  },
  trackDetails: {
    route: '/tracks/:trackId',
    exact: true
  }
};
