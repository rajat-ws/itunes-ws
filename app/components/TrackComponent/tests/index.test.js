/**
 *
 * Tests for TrackComponent
 *
 */

import React from 'react';
import * as routeData from 'react-router-dom';
import { fireEvent } from '@testing-library/dom';
import { renderProvider, renderWithIntl, timeout } from '@utils/testUtils';
import TrackComponent from '../index';
import { MOCK_TRACK_DATA } from '@app/utils/mockData';

describe('<TrackComponent />', () => {
  let handlePauseTrackWrapperSpy = jest.fn();

  it('should render and match the snapshot', () => {
    const { baseElement } = renderWithIntl(
      <TrackComponent trackData={MOCK_TRACK_DATA} handlePauseTrackWrapper={handlePauseTrackWrapperSpy} />
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('should contain 1 TrackComponent component', () => {
    const { getAllByTestId } = renderWithIntl(<TrackComponent trackData={MOCK_TRACK_DATA} />);
    expect(getAllByTestId('track-component').length).toBe(1);
  });

  it('should handle the handleRoute navigation to TrackDetails Container', () => {
    const mockHistory = {
      push: jest.fn()
    };

    const mockHistorySpy = jest.spyOn(routeData, 'useHistory').mockReturnValue(mockHistory);

    const { getByTestId } = renderProvider(<TrackComponent trackData={MOCK_TRACK_DATA} isShowDetailsButton={true} />);
    const button = getByTestId('showDetails');
    expect(button).toHaveTextContent(/Show Details/i);
    fireEvent.click(button);
    expect(mockHistorySpy).toHaveBeenCalled();
  });

  it('should render the PLAY text on PlayTrackBtn button', () => {
    const { getByRole } = renderProvider(
      <TrackComponent trackData={MOCK_TRACK_DATA} handlePauseTrackWrapper={handlePauseTrackWrapperSpy} />
    );
    const button = getByRole('button');

    expect(button).toHaveTextContent(/play/i);
  });

  it('should render the PAUSE text on PlayTrackBtn button when the play is clicked', async () => {
    const { getByRole } = renderProvider(
      <TrackComponent trackData={MOCK_TRACK_DATA} handlePauseTrackWrapper={handlePauseTrackWrapperSpy} />
    );

    // const mockUseRef = {
    //   current: {
    //     play: jest.fn(),
    //     pause: jest.fn()
    //   }
    // };

    // const mockUseRefSpy = jest.spyOn(React, 'useRef').mockReturnValueOnce(mockUseRef);

    const button = getByRole('button');
    expect(button).toHaveTextContent(/play/i);
    fireEvent.click(button);
    await timeout(500);
    expect(button).toHaveTextContent(/pause/i);
    // expect(mockUseRefSpy).toHaveBeenCalled();
  });
});
