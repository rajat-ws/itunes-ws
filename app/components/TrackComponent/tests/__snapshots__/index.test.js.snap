// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TrackComponent /> should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="ant-card ant-card-bordered TrackComponent__TrackCardContainer-sc-1y7fh5t-0 hZfrIv"
      data-testid="track-component"
    >
      <div
        class="ant-card-body"
      >
        <div
          class="ant-image"
          style="width: 80%;"
        >
          <img
            class="ant-image-img"
            src="https://is3-ssl.mzstatic.com/image/thumb/Music124/v4/dd/64/78/dd6478bc-e684-1b69-2e4e-c7e1d6a36c0a/045635953260.jpg/100x100bb.jpg"
          />
          <div
            class="ant-image-mask"
          >
            <div
              class="ant-image-mask-info"
            >
              <span
                aria-label="eye"
                class="anticon anticon-eye"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="eye"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M942.2 486.2C847.4 286.5 704.1 186 512 186c-192.2 0-335.4 100.5-430.2 300.3a60.3 60.3 0 000 51.5C176.6 737.5 319.9 838 512 838c192.2 0 335.4-100.5 430.2-300.3 7.7-16.2 7.7-35 0-51.5zM512 766c-161.3 0-279.4-81.8-362.7-254C232.6 339.8 350.7 258 512 258c161.3 0 279.4 81.8 362.7 254C791.5 684.2 673.4 766 512 766zm-4-430c-97.2 0-176 78.8-176 176s78.8 176 176 176 176-78.8 176-176-78.8-176-176-176zm0 288c-61.9 0-112-50.1-112-112s50.1-112 112-112 112 50.1 112 112-50.1 112-112 112z"
                  />
                </svg>
              </span>
              Preview
            </div>
          </div>
        </div>
        <div
          class="TrackComponent__StyledDescription-sc-1y7fh5t-1 ehlPZI"
        >
          <div
            class="ant-typography"
          >
            <span
              class="TrackComponent__StyledSpan-sc-1y7fh5t-2 jsYJCu"
            >
               Artist name: 
            </span>
             
            Tom MacDonald
          </div>
          <div
            class="ant-typography"
          >
            <span
              class="TrackComponent__StyledSpan-sc-1y7fh5t-2 jsYJCu"
            >
               Collection name: 
            </span>
             
            See You Tomorrow
          </div>
          <div
            class="ant-typography"
          >
            <span
              class="TrackComponent__StyledSpan-sc-1y7fh5t-2 jsYJCu"
            >
               Track name: 
            </span>
             
            Greenday
          </div>
        </div>
        <div
          class="TrackComponent__ButtonWrapper-sc-1y7fh5t-5 dehKAG"
        >
          <button
            class="TrackComponent__PlayTrackBtn-sc-1y7fh5t-3 khRFCr"
          >
            <span
              class="TrackComponent__ButtonLabel-sc-1y7fh5t-6 ftUulf"
            >
               Play 
            </span>
          </button>
        </div>
        <audio
          data-testid="trackAudio"
          src="https://audio-ssl.itunes.apple.com/itunes-assets/AudioPreview125/v4/51/bb/61/51bb6190-6865-3050-af01-0f64502bc376/mzaf_14313183610807291436.plus.aac.p.m4a"
        />
      </div>
    </div>
  </div>
</body>
`;
